src/variable_info_detail.hpp

```diff
@@ -16,80 +16,150 @@
 	{
-		typedef T type;
-	};
+		assert(index >= 0);
+		// cfg.child_or_empty does not support index parameter
+		if(const config& child = cfg.child(key, index)) {
+			return child;
+		}
+
+		static const config empty_const_cfg;
+		return empty_const_cfg;
+	}
 
-	template <class T>
-	struct maybe_const<vit_const, T>
+	static std::string error_message(const std::string& name)
 	{
-		typedef const T type;
-	};
+		return "Cannot resolve variable '" + name + "' for reading.";
+	}
+};
 
-	template <>
-	struct maybe_const<vit_const, config::child_itors>
+/** Creates a child table when resolving name if it doesn't exist yet. */
+class vi_policy_create
+{
+public:
+	static config& get_child_at(config& cfg, const std::string& key, int index)
 	{
-		typedef config::const_child_itors type;
-	};
+		assert(index >= 0);
+		// the 'create_if_not_existent' logic.
+		while(static_cast<int>(cfg.child_count(key)) <= index) {
+			cfg.add_child(key);
+		}
 
+		return cfg.child(key, index);
+	}
 
-	template<const variable_info_type vit>
-	struct variable_info_state
+	static std::string error_message(const std::string& name)
 	{
-		typedef typename maybe_const<vit,config>::type child_t;
-
-		variable_info_state(child_t& vars)
-			: child_(&vars)
-			, key_()
-			, index_(0)
-			, temp_val_()
-			, type_(state_start)
-		{
-			child_ = &vars;
+		return "Cannot resolve variable '" + name + "' for writing.";
+	}
+};
+
+/**
+ * Will throw an exception when trying to access a nonexistent table.
+ * Note that the other types can throw too if name is invlid, such as '..[[[a'.
+ */
+class vi_policy_throw
+{
+public:
+	static config& get_child_at(config& cfg, const std::string& key, int index)
+	{
+		assert(index >= 0);
+		if(config& child = cfg.child(key, index)) {
+			return child;
 		}
 
-		// The meaning of the following 3 depends on 'type_', but usualy the case is:
-		// the current config is  child_->child_at(key_, index_).
-		child_t* child_;
-		std::string key_;
-		int index_;
+		throw invalid_variablename_exception();
+	}
 
-		// If we have a temporary value like .length
-		// Then we store the result here.
-		config::attribute_value temp_val_;
+	static std::string error_message(const std::string& name)
+	{
+		return "Cannot resolve variable '" + name + "' for writing without creating new children.";
+	}
+};
 
-		// See the definition of 'variable_info_state_type'
-		variable_info_state_type type_;
-	};
-}
+// ==================================================================
+// Other implementation details.
+// ==================================================================
+
+template<typename T, typename V>
+struct maybe_const : public utils::const_clone<T, V>
+{
+	// Meta type aliases provided by const_clone
+};
+
+template<>
+struct maybe_const<config::child_itors, const vi_policy_const>
+{
+	using type = config::const_child_itors;
+};
+
+enum variable_info_state_type {
+	state_start = 0, /**< For internal use (TODO: need more info) */
```

Comment:
this is the initial state in case no part of the variable string was processed yet, all other states have a 'last' key (like 'aa' in case of cc.bb.aa) while this one don't need one.